daq/runner.py

```diff
@@ -189,32 +193,35 @@ def _handle_port_state(self, dpid, port, active):
             return
 
         if active != (port in self._active_ports):
-            LOGGER.info('Port %s dpid %s is now active %s', port, dpid, active)
+            LOGGER.info('Port %s dpid %s is now %s', port, dpid, "active" if active else "inactive")
         if active:
-            if not self._active_ports.get(port):
-                self._activate_port(port, True)
+            self._activate_port(port)
+            if port in self._toggling_ports:
+                del self._toggling_ports[port]
         else:
             if port in self.port_targets:
-                self.target_set_complete(port, 'port not active')
+                if port not in self._toggling_ports:
+                    self._toggling_ports[port] = time.time()
             if port in self._active_ports:
-                if self._active_ports[port] is not True:
-                    self._direct_port_traffic(self._active_ports[port], port, None)
-                self._activate_port(port, None)
+                if port in self._mac_port_mapping and port not in self._toggling_ports:
+                    self._direct_port_traffic(self._mac_port_mapping[port], port, None)
+                self._deactivate_port(port)
         self._send_heartbeat()
 
-    def _activate_port(self, port, state):
-        if state:
-            self._active_ports[port] = state
-        elif port in self._active_ports:
-            del self._active_ports[port]
+    def _activate_port(self, port):
+        self._active_ports.add(port)
+
+    def _deactivate_port(self, port):
+        if port in self._active_ports:
+            self._active_ports.remove(port)
 
     def _direct_port_traffic(self, mac, port, target):
         self.network.direct_port_traffic(mac, port, target)
 
     def _handle_port_learn(self, dpid, port, target_mac):
         if self.network.is_device_port(dpid, port):
             LOGGER.info('Port %s dpid %s learned %s', port, dpid, target_mac)
-            self._activate_port(port, target_mac)
+            self._mac_port_mapping[port] = target_mac
```

Comment:
variable name is backwards, it should be _port_mac_mapping (also, see previous comment about combining this with active ports). It might be that the top-level construct should be "port_info" not "active ports", and "is_active" would just be a sub-entry in the top-level dictionary.